OpenProblemLibrary/Hope/Multi1/03-05-Basis-subspace/Basis_30.pg

```diff
@@ -95,7 +95,18 @@ $mc -> extra(
 #$mc -> makeLast("None of the above");
 
 $ans3 = Compute("$a a1 + $c a2");
-$ans4 = Compute("$b a1 + $d a2");
+#$ans4 = Compute("$b a1 + $d a2");
+
+$ans4 = Compute("$b a1 + $d a2")->cmp(
+  checker => sub{
+    my($correct,$student,$ansHash) = @_;
+    # Mark any answer involving explicit nonzero vectors as incorrect
+    if ($student->eval(a1=>$zero,a2=>$zero,a3=>$zero,a4=>$zero) ne $zero) {
+        return 0;
+    }
+    return $a4 == $student->eval(a1=>$a1,a2=>$a2,a3=>$a3, a4=>$zero);
+  }
+);
```

Comment:
This checker should include the following:
```perl
    # a4 should not be allowed in the student formula
    foreach my $var (keys %{$student->{variables}}) {
      Value::Error('You are not allowed to use "a4" in this answer') if $var eq 'a4';
    }
```

Reply:
and I would also recommend using `Value::Error()` to provide feedback in the case where a student answer uses explicit vectors...

Reply:
and L104 should use `==` to test equality with the zero vector, and not `eq` (that's for comparing strings)

Reply:
I see that you changed eval to substitute so that you could drop the `a4` -- but I think it's better to recognize and provide feedback when `a4` is used in an answer.

Reply:
>  and I would also recommend using Value::Error() to provide feedback in the case where a student answer uses explicit vectors...

Thanks for the Value::Error tip. I am wondering what's an appropriate way to detect explicit vectors. Zeroing out the variables, and then comparing with <0,0,0>, still accepts an answer with <0,0,0> in it, which technically should not be allowed in Parts (b) and (c). I guess I could use regex ($student !~ /<|[ijk]/), but perhaps there are better ways? Thanks!

Reply:
Pardon the hiatus. A new update has been made to include error messages for using "forbidden" vectors.